### PR TITLE
Add new Patch2Self CLI script for dMRI denoising

### PR DESCRIPTION
**Pull request by ShreyasFadnavis**
Hi @jcohenadad  and @joshuacwnewton !

Sorry for the slowness on this PR -- Following up with adding Patch2Self in SCT as per discussion in #3234 :)
```
Reference: Fadnavis, et. al., "Patch2Self: Denoising Diffusion MRI data with Self-Supervised Learning", Advances in Neural Information Processing Systems, 2020, Vol. 33.
```

Link: https://papers.nips.cc/paper/2020/hash/bc047286b224b7bfa73d4cb02de1238d-Abstract.html

### CODE REVIEW

**Comment by ShreyasFadnavis** [comment]
As per the discussion in #3234 , the `sct_denoise` was intended to do both dMRI and structural MRI denoising. But I think its much easier to make a separate script for `Patch2Self` as the options are quite different and would not confuse the user. 

@jcohenadad @joshuacwnewton Let me know what you'll think!

IMO: `sct_denoise_onlm` and `sct_denoise_patch2self` could be 2 separate scripts. If you guys think this is not optimal, let me know and I will try to combine the 2 👍🏽

**Comment by joshuacwnewton** [comment]
> As per the discussion in #3234 , the `sct_denoise` was intended to do both dMRI and structural MRI denoising. But I think its much easier to make a separate script for `Patch2Self` as the options are quite different and would not confuse the user.

Ah, you're absolutely right! 

Please disregard the previous comments I made in https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3234. For some reason, at the time I was under the impression that `patch2self` had a similar interface to `nlmeans`, but I was mistaken; as you've pointed out, this is clearly not the case! I hadn't realized that [`patch2self`](https://dipy.org/documentation/latest/reference/dipy.denoise/#dipy.denoise.patch2self.patch2self) was designed to be used specifically with DWI images, while [`nlmeans`](https://dipy.org/documentation/latest/reference/dipy.denoise/#dipy.denoise.nlmeans.nlmeans) can be used on 3D or 4D data arrays. 

> IMO: `sct_denoise_onlm` and `sct_denoise_patch2self` could be 2 separate scripts.

I think this sounds perfectly reasonable. :slightly_smiling_face: 

I do have one question -- you mentioned in https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3234#issuecomment-786762068 that `patch2self` can be used with fMRI data, too. Is this still true? And, if so, how is the `bvals` argument of `patch2self` handled when working with fMRI data? 

(I ask because, if `patch2self` is DWI-only, then prefixing the script with "`sct_dmri`" might be worth considering. But, if it can be used with both dMRI and fMRI data, then it might be better to exclude the "`dmri`" part of the script name.)

**Comment by ShreyasFadnavis** [comment]
Thanks @joshuacwnewton !! I will make the changes according to your suggestions 😄

**Comment by ShreyasFadnavis** [comment]
Okay! I think I have all the components in :) @joshuacwnewton -- needed some help with testing this. Can you test this script on your end on some sample data?

The patch2self function has 2 mandatory arguments:
1. `.nii file`
2. `.bval file`

The rest are optional as I think I have in the code. It's very likely that I have made a mistake as I am not very well-versed with the system. Please let me know!

**Review by joshuacwnewton — CHANGES_REQUESTED** [review]
Thank you so much for the contribution! This is an excellent start. :smile: 

I have a few suggestions, and I also found a few bugs. I'm happy to give another round of reviews once this feedback has been addressed.

File: spinalcordtoolbox/math.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
 from dipy.denoise.noise_estimate import estimate_sigma
 from dipy.segment.mask import median_otsu
 from dipy.denoise.nlmeans import nlmeans
+from dipy.denoise.patch2self import patch2self
```
[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply spinalcordtoolbox/math.py:16]: > I can submit an upstream PR to `dipy` to have this fixed. slightly_smiling_face

Fixed in https://github.com/dipy/dipy/pull/2438. :smile:

File: spinalcordtoolbox/scripts/sct_denoise_patch2self.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    optional.add_argument(
+        "-p",
+        help='Type of regression model used for self-supervised training within Patch2Self.',
+        required=False,
+        choices=("ols", "ridge", 'lasso'),
+        default="ols")
+    optional.add_argument(
+        "-d",
+        type=int,
+        help="Patch Radius used to generate p-neighbourhoods withing Patch2Self",
```
[the remark this answers is not in the document]
Comment (ShreyasFadnavis) [reply spinalcordtoolbox/scripts/sct_denoise_patch2self.py:66]: Yes! You are absolutely right -- fixing this 👍🏽

File: spinalcordtoolbox/scripts/sct_denoise_patch2self.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                     verbose=True)
+
+    axial_middle = int(data.shape[2] / 2)
+    middle_vol = int(data.shape[3] / 2)
+
+    before = data[:, :, axial_middle, middle_vol].T
+    after = den[:, :, axial_middle, middle_vol].T
+
+    diff_4d = np.absolute(den.astype('f8') - data.astype('f8'))
+    difference = np.absolute(after.astype('f8') - before.astype('f8'))
```
[the remark this answers is not in the document]
Comment (ShreyasFadnavis) [reply spinalcordtoolbox/scripts/sct_denoise_patch2self.py:130]: Okay -- will do!

File: spinalcordtoolbox/scripts/sct_denoise_patch2self.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+    optional = parser.add_argument_group("\nOPTIONAL ARGUMENTS")
+    optional.add_argument(
+        "-h",
+        "--help",
+        action="help",
+        help="Show this help message and exit")
+    optional.add_argument(
+        "-p",
+        help='Type of regression model used for self-supervised training within Patch2Self.',
```
[the remark this answers is not in the document]
Comment (ShreyasFadnavis) [reply spinalcordtoolbox/scripts/sct_denoise_patch2self.py:59]: OLS will never fail for Spinal Cord data. But if one was analyzing "big data", I wanted to allow them to use `model='ridge'` or `model='lasso'`!

I will add a help as per your suggestion!

File: spinalcordtoolbox/scripts/sct_denoise_patch2self.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    mandatory = parser.add_argument_group("\nMANDATORY ARGUMENTS")
+    mandatory.add_argument(
+        "-i",
+        default=None,
+        required=True,
+        help="Input NIFTI image to be denoised. Example: image_input.nii.gz",
+        metavar=Metavar.file,
+    )
+
+    mandatory = parser.add_argument_group("\nMANDATORY ARGUMENTS")
```
[the remark this answers is not in the document]
Comment (ShreyasFadnavis) [reply spinalcordtoolbox/scripts/sct_denoise_patch2self.py:42]: Gotcha -- I will combine the 2!

File: spinalcordtoolbox/scripts/sct_denoise_patch2self.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        required=False,
+        choices=("ols", "ridge", 'lasso'),
+        default="ols")
+    optional.add_argument(
+        "-d",
+        type=int,
+        help="Patch Radius used to generate p-neighbourhoods withing Patch2Self",
+        metavar=Metavar.int,
+        required=False,
+        default="[0, 0, 0]")
```
[the remark this answers is not in the document]
Comment (ShreyasFadnavis) [reply spinalcordtoolbox/scripts/sct_denoise_patch2self.py:69]: I will go with option 1 👍🏽

File: spinalcordtoolbox/math.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShreyasFadnavis) [reply spinalcordtoolbox/math.py:16]: @joshuacwnewton : Great point! I have written some tests in DIPY to verify the proper working of Patch2Self:
https://github.com/dipy/dipy/blob/master/dipy/denoise/tests/test_patch2self.py

Do we need to write tests within SCT too?

[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply spinalcordtoolbox/math.py:16]: > Do we need to write tests within SCT too?

That would be much appreciated! 

We have some documentation [in our dev wiki](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Testing) for writing tests. If anything in those pages is unclear, please ask, and I'll be happy to help out.

**Comment by joshuacwnewton** [comment]
Hi @ShreyasFadnavis! I just wanted to follow up on the above PR review. Are you still able to address the feedback that was given?

**Comment by ShreyasFadnavis** [comment]
> Hi @ShreyasFadnavis! I just wanted to follow up on the above PR review. Are you still able to address the feedback that was given?

Hi @joshuacwnewton ! Really sorry for the delay on this. I will get at it early next week!

**Comment by joshuacwnewton** [comment]
> Hi @joshuacwnewton ! Really sorry for the delay on this. I will get at it early next week!

Thank you for letting us know! We understand that things move at a slower pace with open source contributions, so no worries. :relaxed:

**Comment by joshuacwnewton** [comment]
> Hi @joshuacwnewton : Sorry for the slowness on my end! I think it would be faster if you go ahead and finish this PR! Unfortunately, I have been caught up with other unavoidable tasks :(

No worries whatsoever! I would be happy to. Thank you for letting us know! :heart:

**Comment by ShreyasFadnavis** [comment]
Hi @joshuacwnewton -- Looks like the release is going to take much longer. Can we try a simple workaround for now :
`data_test = np.concatenate((data, data), axis=3)` 

Can we use the above `data_test` within SCT for Patch2Self for now? It will speed up things drastically :)

Let me know how I can help!

**Comment by ShreyasFadnavis** [comment]
Hi @joshuacwnewton ! Thank you for these details. I think for the test, as long as we have **more than** 10 volumes, for Patch2Self, it should work perfectly well. 

I did take a look at the SNR computation within SCT (AFAIK: it uses background voxels to estimate the noise level) and I think it would be a good test to add. This maybe not be necessary within SCT as we already do these tests within DIPY :)

I am a little tied up with other things and think that it would much faster if you do it since you know the API of SCT better than me. 

Thanks,
Shreyas

**Comment by joshuacwnewton** [comment]
> I think for the test, as long as we have **more than** 10 volumes, for Patch2Self, it should work perfectly well.

Oh, of course! Thank you for pointing that out. I definitely misspoke when I said <10 volumes. :facepalm: 

> I am a little tied up with other things and think that it would much faster if you do it since you know the API of SCT better than me.

Sounds good! I'll wrap up this PR today, then. :smile:

**Comment by ShreyasFadnavis** [comment]
@joshuacwnewton ! Thanks a lot for pushing this -- All looks green and good now. 1 final request -- We have a preprint on BioArxiv under review: https://www.medrxiv.org/content/10.1101/2021.10.04.21264389v1.full wherein we use SCT pervasively. 

Would it be reasonable to add it to the codebase, or should we wait until publication? Either is good with me just wanted to ask :)

**Comment by jcohenadad** [comment]
> @joshuacwnewton ! Thanks a lot for pushing this -- All looks green and good now. 1 final request -- We have a preprint on BioArxiv under review: https://www.medrxiv.org/content/10.1101/2021.10.04.21264389v1.full wherein we use SCT pervasively.

This is great! Congratulations on this publication! 🎉 
Would it be possible to mention in the publication that the Patch2Self is now available in SCT?

> Would it be reasonable to add it to the codebase, or should we wait until publication? Either is good with me just wanted to ask :)

a preprint _is_ a publication 😊   so yes, we should reference to this publication in the usage of the new CLI, and once the journal publication is out we will replace it

**Comment by ShreyasFadnavis** [comment]
@jcohenadad thank you for your response! Yes, we will add it to the paper ASAP. I have spoken to Dr. Schilling about it too 😃

**Comment by ShreyasFadnavis** [comment]
@schillkg adding you to the conversation here. This PR is almost complete and ready to go!

As per the above conversation, can we specify that P2S is available and disseminated the the spinal cord community via SCT? Perhaps we can also add it in acknowledgments… with a thank you note!

**Comment by schillkg** [comment]
Yes - absolutely! Just got reviews back during the holidays and I can be sure to add in methods and acknowledgements in the next revision.

File: spinalcordtoolbox/scripts/sct_dmri_denoise_patch2self.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+
+def get_parser():
+    parser = SCTArgumentParser(
+        description='Utility function to denoise diffusion MRI images. Return the denoised image and also the difference '
+                    'between the input and the output.The Patch2Self denoising algorithm is based on self-supervised denoising via statistical independence of noise, as described in the following publications:\n'
+                    '\n'
+                    '- Fadnavis et al. Patch2Self: Denoising Diffusion MRI with Self-supervised Learning. NeurIPS, 2020, Vol. 33. (https://arxiv.org/abs/2011.01355)\n'
+                    '- Schilling et al. Patch2Self denoising of diffusion MRI in the cervical spinal cord improves intra-cord contrast, '
+                    'signal modelling, repeatability, and feature conspicuity. Preprint, 2021. (https://doi.org/10.1101/2021.10.04.21264389)\n'
```
Comment (joshuacwnewton) [inline spinalcordtoolbox/scripts/sct_dmri_denoise_patch2self.py:32]: Very sorry for the delay in getting to this PR! I've added a link to the preprint (as suggested by @jcohenadad), which looks like this on my end:

Screenshot from 2022-01-24 12-16-59

File: spinalcordtoolbox/scripts/sct_dmri_denoise_patch2self.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        patch_radius = int(arguments.radius)
+
+    file_to_denoise = arguments.i
+    bval_file = arguments.b
+    output_file_name = arguments.o
+
+    param = Param()
+    param.verbose = verbose
+    param.remove_temp_files = remove_temp_files
+    param.parameter = model
```
Comment (joshuacwnewton) [inline spinalcordtoolbox/scripts/sct_dmri_denoise_patch2self.py:120]: Just reviewing this PR one last time with fresh eyes in the new year. :)

I think that `Param` was something that was copied from our other denoising script, but, looking at its usage, I don't think it's actually necessary here? 

I think we could simplify things by getting rid of `Param` entirely, and sticking only with `arguments` for containing the default arguments.

Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_dmri_denoise_patch2self.py]: Done in c6c1dd0.

File: spinalcordtoolbox/scripts/sct_dmri_denoise_patch2self.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+
+def get_parser():
+    parser = SCTArgumentParser(
+        description='Utility function to denoise diffusion MRI images. Return the denoised image and also the difference '
+                    'between the input and the output.The Patch2Self denoising algorithm is based on self-supervised denoising via statistical independence of noise, as described in the following publications:\n'
+                    '\n'
+                    '- Fadnavis et al. Patch2Self: Denoising Diffusion MRI with Self-supervised Learning. NeurIPS, 2020, Vol. 33. (https://arxiv.org/abs/2011.01355)\n'
+                    '- Schilling et al. Patch2Self denoising of diffusion MRI in the cervical spinal cord improves intra-cord contrast, '
+                    'signal modelling, repeatability, and feature conspicuity. Preprint, 2021. (https://doi.org/10.1101/2021.10.04.21264389)\n'
```
Comment (jcohenadad) [inline spinalcordtoolbox/scripts/sct_dmri_denoise_patch2self.py:19]:
```suggestion
                    '- Schilling et al. Patch2Self denoising of diffusion MRI in the cervical spinal cord improves intra-cord contrast, '
                    'signal modelling, repeatability, and feature conspicuity. medRxiv, 2021. (https://doi.org/10.1101/2021.10.04.21264389)\n'
```

**Review by jcohenadad — APPROVED** [review]

**Comment by ShreyasFadnavis** [comment]
Hi @joshuacwnewton and @jcohenadad ! Thanks for pushing this and sorry for the slowness on my end. All LGTM 👍🏽 

Excited about the release!!

**Comment by joshuacwnewton** [comment]
> Hi @joshuacwnewton and @jcohenadad ! Thanks for pushing this and sorry for the slowness on my end. All LGTM 👍🏽

No worries whatsoever! We really appreciate your contribution nonetheless. Thank you for getting the ball rolling on this, and I'm happy to see this PR has reached its conclusion!  :heart: :tada:

**Pull request merged**